cli/module_registry.go

```diff
@@ -176,46 +174,152 @@ func UpdateModuleCommand(c *cli.Context) error {
 	}
 
 	// If the namespace isn't set, modify the meta.json to set it (if available)
-	if moduleID.Namespace == "" {
+	if moduleID.Prefix == "" {
 		org, err := resolveOrg(client, orgIDArg, publicNamespaceArg)
 		if err != nil {
 			return err
 		}
 		if org.PublicNamespace != "" {
-			moduleID.Namespace = org.PublicNamespace
+			moduleID.Prefix = org.PublicNamespace
 			manifest.Name = moduleID.toString()
 			if err := writeManifest(manifestPath, manifest); err != nil {
 				return err
 			}
 			fmt.Fprintf(c.App.Writer, "\nUpdated %s to use the public namespace of %q which is %q\n",
-				manifestPath, org.Name, org.PublicNamespace)
+				defaultManifestFilename, org.Name, org.PublicNamespace)
 			fmt.Fprintf(c.App.Writer, "You no longer need to specify org_id or public_namespace\n")
 		}
 	}
 
 	return nil
 }
 
+// UploadModuleCommand runs the command to upload a new version of a module.
+func UploadModuleCommand(c *cli.Context) error {
+	manifestPathArg := c.String("module")
+	publicNamespaceArg := c.String("public_namespace")
+	orgIDArg := c.String("org_id")
+	nameArg := c.String("name")
+	versionArg := c.String("version")
+	platformArg := c.String("platform")
+	tarballPath := c.Args().First()
+	if c.Args().Len() > 1 {
+		return errors.New("Too many arguments passed to upload command. " +
+			"Make sure to specify flag and optional arguments before the required positional package argument")
+	}
+	if tarballPath == "" {
+		return errors.New("No package to upload -- please provide an archive containing your module. See the help for more information")
+	}
+
+	client, err := NewAppClient(c)
+	if err != nil {
+		return err
+	}
+
+	manifestPath := defaultManifestFilename
+	if manifestPathArg != "" {
+		manifestPath = manifestPathArg
+	}
+	var moduleID ModuleID
+	// if the manifest cant be found
+	if _, err := os.Stat(manifestPath); err != nil {
+		// no manifest found.
+		if nameArg == "" || (publicNamespaceArg == "" && orgIDArg == "") {
+			return errors.Errorf("Unable to find %s. "+
+				"If you want to upload a version without a %s, you must supply a module name and namespace (or module name and orgid)\n\n",
+				defaultManifestFilename, defaultManifestFilename)
+		}
+		moduleID = ModuleID{Prefix: publicNamespaceArg, Name: nameArg}
+	} else {
+		// if we can find a manifest, use that
+		manifest, err := loadManifest(manifestPath)
+		if err != nil {
+			return err
+		}
+
+		moduleID, err = parseModuleID(manifest.Name)
+		if err != nil {
+			return err
+		}
+		if nameArg != "" && nameArg != moduleID.Name {
+			// This is almost certainly a mistake we want to catch
+			return errors.Errorf("Module name %q was supplied via command line args but the %s has a module name of %q",
```

Comment:
nit: maybe indicate that this is named in the manifest file so its clear.

Reply:
For some reason I had been avoiding hardcoding `defaultManifestFIlename` in these strings and had been string formatting them, but your message has convinced me that it is not worth it ans is confusing. 

All error messages will now have `meta.json` hardcoded in place of `defaultManifestFilename`